### PR TITLE
Add snippets for most Ruby constructs

### PR DESCRIPTION
**Pull request by jez**
### Motivation

If people use these snippets, they'll take the fast path more often because
there will be fewer parse errors. We already recover from many kinds of parse
errors, but not all of them, and not always as gracefully as we'd like. So flat
out avoiding parse errors in the first place should also help out.

### Test plan

See included automated tests.

### CODE REVIEW

File: main/lsp/requests/completion.cc [hunk 1]
```diff
@@ -18,43 +18,50 @@ namespace {
...
-    {"def", "Defines a method."},
-    {"defined?", "Returns a string describing its argument."},
-    {"do", "Starts a block."},
-    {"else", "The unhandled condition in case, if and unless expressions."},
-    {"elsif", "An alternate condition for an if expression."},
+    {"case", "Starts a case expression.", "case ${1:expr}\nwhen ${2:expr}\n  $0\nelse\nend"},
+    {"class", "Creates or opens a class.", "class ${1:ClassName}\n  $0\nend"},
+    {"def", "Defines a method.", "def ${1:method_name}($2)\n  $0\nend"},
+    {"defined?", "Returns a string describing its argument.", "defined?(${1:Constant})$0"},
+    // TODO(jez) Even better would be to auto-insert a block for methods that we know must take a block
```
Comment (elliottt) [inline main/lsp/requests/completion.cc]: This would be pretty cool, is that something that's on the roadmap?

File: main/lsp/requests/completion.cc [hunk 2]
```diff
@@ -63,10 +70,11 @@ const RubyKeyword rubyKeywords[] = {
     {"true", "Boolean true."},
     // This is also defined on Kernel
     // {"undef", "Prevents a class or module from responding to a method call."},
-    {"unless", "Used for unless and modifier unless expressions."},
-    {"until", "Creates a loop that executes until the condition is true."},
-    {"when", "A condition in a case expression."},
-    {"while", "Creates a loop that executes while the condition is true."},
+    {"unless", "Used for unless and modifier unless expressions.", "unless ${1:expr}\n  $0\nend"},
+    {"until", "Creates a loop that executes until the condition is true.", "until ${1:expr}\n  $0\nend"},
+    // Would really like to dedent the line too...
```
Comment (elliottt) [inline main/lsp/requests/completion.cc]: Is that something that can be expressed with the snippets?

File: main/lsp/requests/completion.cc [hunk 3]
```diff
@@ -63,10 +70,11 @@ const RubyKeyword rubyKeywords[] = {
     {"true", "Boolean true."},
     // This is also defined on Kernel
     // {"undef", "Prevents a class or module from responding to a method call."},
-    {"unless", "Used for unless and modifier unless expressions."},
-    {"until", "Creates a loop that executes until the condition is true."},
-    {"when", "A condition in a case expression."},
-    {"while", "Creates a loop that executes while the condition is true."},
+    {"unless", "Used for unless and modifier unless expressions.", "unless ${1:expr}\n  $0\nend"},
```
Comment (elliottt) [inline main/lsp/requests/completion.cc]: Can this one fire if you're writing a trailing `unless`? Can this:
```
x = 10 unless
```
expand out to this:
```
x = 10 unless ${1:expr}
end
```

File: main/lsp/requests/completion.cc [hunk 4]
```diff
@@ -254,27 +262,39 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method, core::
...
+    }
+
     auto replacementLoc = core::Loc{queryLoc.file(), queryStart - prefixSize, queryStart};
     auto replacementRange = Range::fromLoc(gs, replacementLoc);
-    auto replacementText = label;
     if (replacementRange != nullptr) {
         item->textEdit = make_unique<TextEdit>(std::move(replacementRange), replacementText);
     } else {
-        // TODO(jez) Why is replacementRange nullptr? instrument this and investigate when it fails
+        // Range::fromLoc failed... maybe the fuzzer is running?
```
Comment (elliottt) [inline main/lsp/requests/completion.cc]: Thanks for updating this TODO!

File: main/lsp/requests/completion.cc [hunk 5]
```diff
@@ -254,27 +262,39 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method, core::
...
         item->insertText = replacementText;
     }
-    item->insertTextFormat = InsertTextFormat::PlainText;
 
-    item->documentation =
-        make_unique<MarkupContent>(config.getClientConfig().clientCompletionItemMarkupKind, rubyKeyword.documentation);
+    item->detail = fmt::format("(sorbet) {}", rubyKeyword.documentation);
+    if (rubyKeyword.snippet.has_value()) {
```
Comment (elliottt) [inline main/lsp/requests/completion.cc]: Should this be guarded on `config.getClientConfig().clientCompletionItemSnippetSupport` as well?

File: main/lsp/requests/completion.cc (hunk 1 above)
Comment (jvilk-stripe) [reply main/lsp/requests/completion.cc]: Could you open a GitHub issue for this? :)

File: main/lsp/requests/completion.cc [hunk 6]
```diff
@@ -18,43 +18,50 @@ namespace {
...
-    {"case", "Starts a case expression."},
-    {"class", "Creates or opens a class."},
-    {"def", "Defines a method."},
-    {"defined?", "Returns a string describing its argument."},
-    {"do", "Starts a block."},
-    {"else", "The unhandled condition in case, if and unless expressions."},
-    {"elsif", "An alternate condition for an if expression."},
+    {"case", "Starts a case expression.", "case ${1:expr}\nwhen ${2:expr}\n  $0\nelse\nend"},
+    {"class", "Creates or opens a class.", "class ${1:ClassName}\n  $0\nend"},
+    {"def", "Defines a method.", "def ${1:method_name}($2)\n  $0\nend"},
```
Comment (jvilk-stripe) [inline main/lsp/requests/completion.cc]: Is it worth naming `$2`?

File: main/lsp/requests/completion.cc [hunk 7]
```diff
@@ -18,43 +18,50 @@ namespace {
...
+    {"begin", "Starts an exception handling block.", "begin\n  $0\nend"},
     {"break", "Leaves a block early."},
-    {"case", "Starts a case expression."},
-    {"class", "Creates or opens a class."},
-    {"def", "Defines a method."},
-    {"defined?", "Returns a string describing its argument."},
-    {"do", "Starts a block."},
-    {"else", "The unhandled condition in case, if and unless expressions."},
-    {"elsif", "An alternate condition for an if expression."},
+    {"case", "Starts a case expression.", "case ${1:expr}\nwhen ${2:expr}\n  $0\nelse\nend"},
```
Comment (jvilk-stripe) [inline main/lsp/requests/completion.cc]: Does this work as expected?

> Placeholders with equal identifiers are linked, that is typing in one will update others too

Are identifiers equal if the names are the same but numbers are different? (Maybe not.)

File: main/lsp/requests/completion.cc [hunk 8]
```diff
@@ -18,43 +18,50 @@ namespace {
...
+    {"class", "Creates or opens a class.", "class ${1:ClassName}\n  $0\nend"},
+    {"def", "Defines a method.", "def ${1:method_name}($2)\n  $0\nend"},
+    {"defined?", "Returns a string describing its argument.", "defined?(${1:Constant})$0"},
+    // TODO(jez) Even better would be to auto-insert a block for methods that we know must take a block
+    {"do", "Starts a block.", "do\n  $0\nend"},
+    {"else", "The unhandled condition in case, if and unless expressions.", "else ${1:expr}$0"},
```
Comment (jvilk-stripe) [inline main/lsp/requests/completion.cc]: Should this just be `else $0`? There's no expression to evaluate as part of the conditional. (Unless this is handy when using an if/else's rval and the else is just one statement...?)

File: main/lsp/requests/completion.cc [hunk 9]
```diff
@@ -18,43 +18,50 @@ namespace {
...
+    {"elsif", "An alternate condition for an if expression.", "elsif ${1:expr}$0"},
     {"end",
      "The end of a syntax block. Used by classes, modules, methods, exception handling and control expressions."},
     {"ensure", "Starts a section of code that is always run when an exception is raised."},
     {"false", "Boolean false."},
     {"for", "A loop that is similar to using the each method."},
```
Comment (jvilk-stripe) [inline main/lsp/requests/completion.cc]: Is there no good snippet for `for`? (`"for ${1:expr} in ${2:expr}\n  $0\nend"` ?

File: main/lsp/requests/completion.cc [hunk 10]
```diff
@@ -18,43 +18,50 @@ namespace {
...
-    {"if", "Used for if and modifier if expressions."},
+    {"if", "Used for if and modifier if expressions.", "if ${1:expr}\n  $0\nend"},
```
Comment (jvilk-stripe) [inline main/lsp/requests/completion.cc]: @elliottt 's comment about `unless` applies to `if` too.

File: main/lsp/requests/completion.cc (hunk 2 above)
Comment (jvilk-stripe) [reply main/lsp/requests/completion.cc]: I don't think so. This is probably something that'll have to wait for autoformat.

File: main/lsp/requests/completion.cc [hunk 11]
```diff
@@ -254,27 +262,39 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method, core::
...
     } else {
-        // TODO(jez) Why is replacementRange nullptr? instrument this and investigate when it fails
+        // Range::fromLoc failed... maybe the fuzzer is running?
         item->insertText = replacementText;
     }
-    item->insertTextFormat = InsertTextFormat::PlainText;
 
-    item->documentation =
-        make_unique<MarkupContent>(config.getClientConfig().clientCompletionItemMarkupKind, rubyKeyword.documentation);
+    item->detail = fmt::format("(sorbet) {}", rubyKeyword.documentation);
```
Comment (jvilk-stripe) [inline main/lsp/requests/completion.cc]: If you want to be cute... we could probably use the 🍦emoji. https://emojipedia.org/soft-ice-cream/ or 🍨https://emojipedia.org/ice-cream/

Dunno if vim / emacs would choke on it, though? (But, then again, that'd mean they'd choke on unicode completion items.)

File: main/lsp/requests/completion.cc (hunk 7 above)
Comment (jez) [reply main/lsp/requests/completion.cc]: oh cool TIL

yeah, if the number is different, they're not linked, even if they have the same placeholder text. also, if the number is the same they **are** linked, even if the placeholder text is different.

tl;dr: this snippet works as written in the PR. thanks!

File: main/lsp/requests/completion.cc (hunk 8 above)
Comment (jez) [reply main/lsp/requests/completion.cc]: ah yes, thanks! copy / paste error

File: main/lsp/requests/completion.cc (hunk 9 above)
Comment (jez) [reply main/lsp/requests/completion.cc]: Ah maybe; i just know that we lint against `for` in pay-server so i didn't want to think about what the syntax looked like.

File: main/lsp/requests/completion.cc (hunk 3 above)
Comment (jez) [reply main/lsp/requests/completion.cc]: Yeah, i punted on `if` vs `post-if` and same for unless... i'm not sure of an easy way around this. I think that people will just learn to not take the snippet

one alternative is to have two snippets, one for post + one for pre, and let the user choose

File: main/lsp/requests/completion.cc (hunk 1 above)
Comment (jez) [reply main/lsp/requests/completion.cc]: for now this is on my list (paper doc) of polish things to add

i plan on making anything that is still in my paper doc into an issue at the point where i call completion "done" and roll onto some other workstream

File: main/lsp/requests/completion.cc (hunk 6 above)
Comment (jez) [reply main/lsp/requests/completion.cc]: when i tried this out, it felt better to just have the tabstop there and not have a default placeholder

File: main/lsp/requests/completion.cc (hunk 11 above)
Comment (jez) [reply main/lsp/requests/completion.cc]: also i don't think that either of those are particularly obviously "Sorbet"

i think vim / emacs would be fine, but i think this is just slightly more explicit.

File: main/lsp/requests/completion.cc (hunk 3 above)
Comment (jez) [reply main/lsp/requests/completion.cc]: added this to my list.
